sdk/cosmos/azure-cosmos/azure/cosmos/_service_request_retry_policy.py

```diff
@@ -43,28 +45,85 @@ def ShouldRetry(self):
             if self.request.operation_type == ResourceType.DatabaseAccount:
                 return False
 
-            if self.location_endpoint:
-                if _OperationType.IsReadOnlyOperation(self.request.operation_type):
-                    self.global_endpoint_manager.mark_endpoint_unavailable_for_read(self.location_endpoint)
-                    self.logger.warning("Marking %s unavailable for read", self.location_endpoint)
-                else:
-                    self.global_endpoint_manager.mark_endpoint_unavailable_for_write(self.location_endpoint)
-                    self.logger.warning("Marking %s unavailable for write", self.location_endpoint)
+            refresh_cache = self.request.last_routed_location_endpoint_within_region is not None
+            self.mark_endpoint_unavailable(refresh_cache)
 
-        self.failover_retry_count += 1
-        if self.failover_retry_count >= self.total_retries:
-            return False
-
-        if self.request:
-            # clear previous location-based routing directive
-            self.request.clear_route_to_location()
+            # Check if it is safe to do another retry
+            if self.in_region_retry_count >= self.total_in_region_retries:
+                # reset the in region retry count
+                self.in_region_retry_count = 0
+                self.failover_retry_count += 1
+                if self.failover_retry_count >= self.total_retries:
+                    return False
 
-            # set location-based routing directive based on retry count
-            # ensuring usePreferredLocations is set to True for retry
-            self.request.route_to_location_with_preferred_location_flag(self.failover_retry_count, True)
+            self.request.last_routed_location_endpoint_within_region = self.request.location_endpoint_to_route
+            if _OperationType.IsReadOnlyOperation(self.request.operation_type):
+                # We just directly got to the next location in case of read requests
+                # We don't retry again on the same region for regional endpoint
+                self.failover_retry_count += 1
+                if self.failover_retry_count >= self.total_retries:
+                    return False
+                # # Check if it is safe to failover to another region
+                self.location_endpoint = self.resolve_next_region_service_endpoint()
+            else:
+                self.location_endpoint = self.resolve_current_region_service_endpoint()
```

Comment:
L69 basically means if another endpoint is resolvable in the same region for writes then flow of control won't enter L72 at all.

Reply:
Is there a way to check if write retries don't get stuck in the same region w/o exiting at all.

Reply:
Every time we retry in the current region, we increment the `in_region_retry_count`, which should protect against any endless loop, since total max in region retries allowed are 2.